Kernel/Storage/SD/SDHostController.cpp

```diff
@@ -0,0 +1,667 @@
+        timeout--;
+    }
+    return timeout > 0;
+}
+
+ErrorOr<void> SDHostController::issue_command(SD::CommandIndex index,
+    u32 argument)
+{
+    // SDHC 3.7.1 Transaction Control without Data Transfer Using DAT Line
+    constexpr u32 command_inhibit = 1 << 1;
+    auto cmd = SD::get_command(index);
+
+    // 1. Check Command Inhibit (CMD) in the Present State register.
+    //    Repeat this step until **Command Inhibit (CMD)** is 0.
+    //    That is, when Command Inhibit (CMD) is 1, the Host Driver
+    //    shall not issue an SD Command.
+    if (!retry_with_timeout(
+            [&]() { return !(m_registers->present_state & command_inhibit); })) {
+        return EIO;
+    }
+
+    // 2. If the Host Driver issues an SD Command using DAT lines
+    //    including busy signal, go to step (3).
+    //    If without using DAT lines including busy signal, go to step (5).
+    // 3. If the Host Driver is issuing an abort command, go to step (5). In the
+    // case of non-abort
+    //    command, go to step (4).
+    if (cmd.requires_dat_line() && cmd.type != SD::CommandType::Abort) {
+
+        // 4. Check Command Inhibit (DAT) in the Present State register. Repeat
+        // this step until
+        //    Command Inhibit (DAT) is set to 0.
+        constexpr u32 data_inhibit = 1 << 2;
+        if (!retry_with_timeout(
+                [&]() { return !(m_registers->present_state & data_inhibit); })) {
+            return EIO;
+        }
+    }
+
+    // 5. Set registers as described in Table 1-2 except Command register.
+    m_registers->argument_1 = argument;
+
+    // 6. Set the Command register.
+    m_registers->transfer_mode_and_command = cmd.raw;
+
+    // 7. Perform Command Completion Sequence in accordance with 3.7.1.2.
+    // Done in wait_for_response()
+
+    return {};
+}
+
+ErrorOr<SDHostController::Response> SDHostController::wait_for_response()
+{
+    // SDHC 3.7.1.2 The Sequence to Finalize a Command
+
+    // 1. Wait for the Command Complete Interrupt. If the Command Complete
+    // Interrupt has occurred,
+    //    go to step (2).
+    if (!retry_with_timeout(
+            [&]() { return m_registers->interrupt_status & command_complete; })) {
+        return EIO;
+    }
+
+    // 2. Write 1 to Command Complete in the Normal Interrupt Status register to
+    // clear this bit
+    m_registers->interrupt_status = command_complete;
+
+    // 3. Read the Response register(s) to get the response.
+    Response r = {};
+    auto cmd = last_sent_command();
+    switch (cmd.response_type) {
+    case SD::ResponseType::NoResponse:
+        break;
+    case SD::ResponseType::ResponseOf136Bits:
+        r.response[0] = m_registers->response_0;
+        r.response[1] = m_registers->response_1;
+        r.response[2] = m_registers->response_2;
+        r.response[3] = m_registers->response_3;
+        break;
+    case SD::ResponseType::ResponseOf48Bits:
+        r.response[0] = m_registers->response_0;
```

Comment:
Correct me if I am wrong, but the bit numbers dont seem to add up
in the first case we read 128 out of 136 (Delta of 8 bit) and in the second 32 out of 48 (Delta of 16 bit)

Maybe add a omment to explain the discrepancies

Reply:
That's because there are technically some extra bits in the physical communication layer, stuff like start&end bits and a checksum. We don't receive any of that stuff at our level of abstraction though.

I have added a comment explaining this.